// @auth0/nextjs-auth0/dist/auth0-session/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientFactory = exports.callbackHandler = exports.logoutHandler = exports.loginHandler = exports.getConfig = exports.TransientStore = exports.StatefulSession = exports.AbstractSession = exports.StatelessSession = exports.ApplicationError = exports.IdentityProviderError = exports.MissingStateCookieError = exports.MissingStateParamError = exports.Cookies = exports.NodeCookies = void 0;
var cookies_1 = require("./utils/cookies");
Object.defineProperty(exports, "NodeCookies", { enumerable: true, get: function () { return __importDefault(cookies_1).default; } });
Object.defineProperty(exports, "Cookies", { enumerable: true, get: function () { return cookies_1.Cookies; } });
var errors_1 = require("./utils/errors");
Object.defineProperty(exports, "MissingStateParamError", { enumerable: true, get: function () { return errors_1.MissingStateParamError; } });
Object.defineProperty(exports, "MissingStateCookieError", { enumerable: true, get: function () { return errors_1.MissingStateCookieError; } });
Object.defineProperty(exports, "IdentityProviderError", { enumerable: true, get: function () { return errors_1.IdentityProviderError; } });
Object.defineProperty(exports, "ApplicationError", { enumerable: true, get: function () { return errors_1.ApplicationError; } });
var stateless_session_1 = require("./session/stateless-session");
Object.defineProperty(exports, "StatelessSession", { enumerable: true, get: function () { return stateless_session_1.StatelessSession; } });
var abstract_session_1 = require("./session/abstract-session");
Object.defineProperty(exports, "AbstractSession", { enumerable: true, get: function () { return abstract_session_1.AbstractSession; } });
var stateful_session_1 = require("./session/stateful-session");
Object.defineProperty(exports, "StatefulSession", { enumerable: true, get: function () { return stateful_session_1.StatefulSession; } });
var transient_store_1 = require("./transient-store");
Object.defineProperty(exports, "TransientStore", { enumerable: true, get: function () { return __importDefault(transient_store_1).default; } });
var get_config_1 = require("./get-config");
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return get_config_1.get; } });
var login_1 = require("./handlers/login");
Object.defineProperty(exports, "loginHandler", { enumerable: true, get: function () { return __importDefault(login_1).default; } });
var logout_1 = require("./handlers/logout");
Object.defineProperty(exports, "logoutHandler", { enumerable: true, get: function () { return __importDefault(logout_1).default; } });
var callback_1 = require("./handlers/callback");
Object.defineProperty(exports, "callbackHandler", { enumerable: true, get: function () { return __importDefault(callback_1).default; } });
var client_1 = require("./client");
Object.defineProperty(exports, "clientFactory", { enumerable: true, get: function () { return __importDefault(client_1).default; } });
